File: Introduction to Robotics Course/Chapter19/Lesson6/Chapter19_Lesson6.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class RobotStats {
    public static void main(String[] args) throws IOException {
        Map<Integer, List<Double>> data = new HashMap<>();

        try (BufferedReader br = new BufferedReader(new FileReader("robot_logs.csv"))) {
            String line = br.readLine(); // header
            while ((line = br.readLine()) != null) {
                String[] tokens = line.split(",");
                double yRef = Double.parseDouble(tokens[1]);
                double yMeas = Double.parseDouble(tokens[2]);
                int trialId = Integer.parseInt(tokens[4]);
                double e = yRef - yMeas;

                data.computeIfAbsent(trialId, k -> new ArrayList<>()).add(e);
            }
        }

        List<Double> Jvalues = new ArrayList<>();
        for (Map.Entry<Integer, List<Double>> entry : data.entrySet()) {
            int trialId = entry.getKey();
            List<Double> samples = entry.getValue();
            double sumSq = 0.0;
            for (double e : samples) {
                sumSq += e * e;
            }
            double J = Math.sqrt(sumSq / samples.size());
            Jvalues.add(J);
            System.out.println("Trial " + trialId + " RMS error = " + J);
        }

        int N = Jvalues.size();
        double mean = 0.0;
        for (double v : Jvalues) mean += v;
        mean /= N;

        double var = 0.0;
        for (double v : Jvalues) {
            double d = v - mean;
            var += d * d;
        }
        var /= (N - 1);
        double stddev = Math.sqrt(var);

        System.out.println("Mean RMS error = " + mean);
        System.out.println("Std of RMS error = " + stddev);
    }
}
      

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter19/Lesson6/Chapter19_Lesson6.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <cmath>

struct Sample {
    double e;
};

int main() {
    std::ifstream in("robot_logs.csv");
    if (!in) {
        std::cerr << "Cannot open file\n";
        return 1;
    }

    std::string header;
    std::getline(in, header); // skip header

    // Map trial_id to list of error samples
    std::unordered_map<int, std::vector<Sample>> data;

    std::string line;
    while (std::getline(in, line)) {
        std::stringstream ss(line);
        std::string token;
        double time, y_ref, y_meas, u;
        int trial_id;

        std::getline(ss, token, ','); time = std::stod(token);
        std::getline(ss, token, ','); y_ref = std::stod(token);
        std::getline(ss, token, ','); y_meas = std::stod(token);
        std::getline(ss, token, ','); u = std::stod(token);
        std::getline(ss, token, ','); trial_id = std::stoi(token);

        Sample s;
        s.e = y_ref - y_meas;
        data[trial_id].push_back(s);
    }

    std::vector<double> J_values;
    for (auto &kv : data) {
        const auto &samples = kv.second;
        double sum_sq = 0.0;
        for (const auto &s : samples) {
            sum_sq += s.e * s.e;
        }
        double J_rms = std::sqrt(sum_sq / samples.size());
        J_values.push_back(J_rms);
        std::cout << "Trial " << kv.first << " RMS error = " << J_rms << "\n";
    }

    int N = static_cast<int>(J_values.size());
    double mean = 0.0;
    for (double v : J_values) mean += v;
    mean /= N;

    double var = 0.0;
    for (double v : J_values) {
        double d = v - mean;
        var += d * d;
    }
    var /= (N - 1);
    double stddev = std::sqrt(var);

    std::cout << "Mean RMS error = " << mean << "\n";
    std::cout << "Std of RMS error = " << stddev << "\n";

    return 0;
}
      
